File: ios/Managers/KeepAwakeManager.swift
class KeepAwakeManager {
    static let shared = KeepAwakeManager()

    private lazy var expoViews = MediaPlayerManager.shared.expoViews

    private func activateKeepAwake() {
        DispatchQueue.main.async {
            UIApplication.shared.isIdleTimerDisabled = true
        }
    }

    private func deactivateKeepAwake() {
        DispatchQueue.main.async {
            UIApplication.shared.isIdleTimerDisabled = false
        }
    }

    func toggleKeepAwake() {
        let anyPlayingView = expoViews.allObjects.contains { view in
            view.mediaPlayer?.isPlaying == true
        }

        if anyPlayingView {
            activateKeepAwake()
        } else {
            deactivateKeepAwake()
        }
    }
}

File: android/src/main/java/expo/modules/libvlcplayer/managers/KeepAwakeManager.kt
package expo.modules.libvlcplayer.managers

import android.app.Activity
import android.view.WindowManager
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.exception.Exceptions

class KeepAwakeManager(
    private val appContext: AppContext,
) {
    private val activity: Activity
        get() = appContext.currentActivity ?: throw Exceptions.MissingActivity()

    private val anyPlayingView: Boolean
        get() =
            MediaPlayerManager.expoViews.any { view ->
                view.mediaPlayer?.isPlaying() == true
            }

    fun activateKeepAwake() {
        activity.let { activity ->
            activity.runOnUiThread {
                activity.window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
            }
        }
    }

    fun deactivateKeepAwake() {
        activity.let { activity ->
            activity.runOnUiThread {
                activity.window.clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
            }
        }
    }

    fun toggleKeepAwake() {
        if (anyPlayingView) {
            activateKeepAwake()
        } else {
            deactivateKeepAwake()
        }
    }
}
